lib/wibox/layout/align.lua

```diff
@@ -35,128 +34,391 @@ local base = require("wibox.widget.base")
-        local w, h = base.fit_widget(self, context, self._private.second, width, height)
-        size_second = self._private.dir == "y" and h or w
-        -- if all the space is taken, skip the rest, and draw just the middle
-        -- widget
-        if size_second >= size_remains then
-            return { base.place_widget_at(self._private.second, 0, 0, width, height) }
-        else
-            -- the middle widget is sized first, the outside widgets are given
-            --  the remaining space if available we will draw later
-            size_remains = floor((size_remains - size_second) / 2)
+    -- Wrapper around layout_place_widget function to pass less parameters
+    local function prepare_place_widget(widget, offset, size, force)
+        -- Let widgets with size of 0 be placed if specified
+        local min_size = force and -1 or 0
+        -- Can't place a widget if it does not exist or the size is invalid
+        if not widget or size <= min_size then
+            return nil
         end
+
+        return layout_prepare_place_widget(
+            self, width, height,
+            widget, offset, size
+        )
     end
-    if self._private.first then
-        local w, h, _ = width, height, nil
-        -- we use the fit function for the "inside" and "none" modes, but
-        --  ignore it for the "outside" mode, which will force it to expand
-        --  into the remaining space
-        if self._private.expand ~= "outside" then
-            if self._private.dir == "y" then
-                _, h = base.fit_widget(self, context, self._private.first, width, size_remains)
-                size_first = h
-                -- for "inside", the third widget will get a chance to use the
-                --  remaining space, then the middle widget. For "none" we give
-                --  the third widget the remaining space if there was no second
-                --  widget to take up any space (as the first if block is skipped
-                --  if this is the case)
-                if self._private.expand == "inside" or not self._private.second then
-                    size_remains = size_remains - h
-                end
-            else
-                w, _ = base.fit_widget(self, context, self._private.first, size_remains, height)
-                size_first = w
-                if self._private.expand == "inside" or not self._private.second then
-                    size_remains = size_remains - w
-                end
-            end
-        else
-            if self._private.dir == "y" then
-                h = size_remains
-            else
-                w = size_remains
-            end
+
+    local function filter_nil(t)
+        local filtered = {}
+        for _, v in pairs(t) do
+            filtered[#filtered + 1] = v
         end
-        table.insert(result, base.place_widget_at(self._private.first, 0, 0, w, h))
+
+        return filtered
```

Comment:
I corrected it. This also fixed the bug with widgets not getting placed when not all 3 slots were populated, since `pairs` would stop at the first `nil` it encounters.

Reply:
> `pairs` would stop at the first `nil` it encounters.

It doesn't. As I said, `pairs` has no notion of "order" in table keys, so it cannot know a "first `nil`". `pairs` always iterates over all present fields, both string and numeric indices.
`ipairs` is the one that is defined as "iterate over integer indices starting from `1` in ascending order, _up to the first empty one_".

Reply:
This makes sense, thanks for the explanation.
This code was removed and now I use `gears.table.from_sparse` instead of this function.